DevelopGitHub/GitHubExtension/GitHubExtension.Activity.External.WebAPI/Services/GitHubService.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using GitHubExtension.Activity.External.WebAPI.Exceptions;
+using GitHubExtension.Security.WebApi.Library.Exceptions;
+using Newtonsoft.Json.Linq;
+
+namespace GitHubExtension.Activity.External.WebAPI.Services
+{
+    public class GitHubService : IGitHubService
+    {
+        private const string LastLinkRelRegex = "last";
+        private const string LinklRelPageRegex = @"page=(\d+)";
+        private readonly HttpClient _httpClient;
+        private IEnumerable<string> _lastRequestLinkHeader;
+ 
+        private static readonly Dictionary<string, string> DefaultHeaders = new Dictionary<string, string>()
+        {
+             //Need to set user-agent to access GitHub API, Using Chrome 48
+            { "User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/48.0.2564.116 Safari/537.36" }
+        };
+
+        private const string EventsUrl = "https://api.github.com/repos/{0}/events?access_token={1}&page={2}";
+
+        public GitHubService()
+        {
+            _httpClient = new HttpClient();
+        }
+
+        public async Task<IEnumerable<JToken>> GetGitHubEventsAsync(string fullRepositoryName, string token, int page)
+        {
+            var message = CreateMessage(HttpMethod.Get, string.Format(EventsUrl, fullRepositoryName, token, page));
+            HttpResponseMessage response = await _httpClient.SendAsync(message);
+
+            if (!response.IsSuccessStatusCode)
+                throw new UnsuccessfullGitHubRequestException();
+            response.Headers.TryGetValues("Link", out _lastRequestLinkHeader);
+
+            JArray events = JArray.Parse(await response.Content.ReadAsStringAsync());
+
+            var parsedEvents = events.Children().Select(e => e.ParseEvent());
+
+            return parsedEvents;
+        }
+
+        /// <summary>
+        /// Used for pagination, gets number of last page from link header of last request
+        /// </summary>
+        /// <returns> int number of pages or null if we are on last page or there are no link </returns>
+        public int? GetNumberOfPages()
+        {
+            if (_lastRequestLinkHeader == null)
+                return null;
+
+            string link = _lastRequestLinkHeader.FirstOrDefault();
+            if (link == null)
+                throw new LinkHeaderMissingException();
+
+            string[][] parts = link.Split(',').Select(l => l.Split(';')).ToArray();
```

Comment:
This code unreadable...I don't want convert in my mind all this stuff. Please do refactoring and extract logic to small methods with a meaningful names.